Extract DeviceType to a standalone header file

desertfire opened the pull request:
Summary:
The goal of this PR and future follow-up PRs is to group a set of header files required by AOTInductor Standalone in a separate directory, ensuring they are implemented in a header-only manner. More specifically, here is what this PR does:
* Extract the DeviceType enum class into a standalone header file in a new torch/standalone/header_only directory
* Retain the existing c10/core/DeviceType.[h|cpp] files to handle complex logic and static variables
* Import symbols from the new torch::standalone namespace into c10 for backward compatibility

This is an updated version of https://github.com/pytorch/pytorch/pull/152787, because we need to land in fbcode first. See the original comments and discussions in https://github.com/pytorch/pytorch/pull/152787.

Test Plan: CI

Differential Revision: D75313723




cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

desertfire commented on torch/standalone/core/DeviceType.h:
```diff
@@ -0,0 +1,221 @@
+  }
+}
+
+#ifdef TORCH_STANDALONE
```
@janeyx99 , @albanD , this is what different from the previous version, which is a good compromise between header-only and standalone needs.

janeyx99 reviewed (approved):
I'm fine with a flag called TORCH_STANDALONE

desertfire committed:
Extract DeviceType to a standalone header file (#154417)

Summary:
X-link: https://github.com/pytorch/executorch/pull/11145

Pull Request resolved: https://github.com/pytorch/pytorch/pull/154417

The goal of this PR and future follow-up PRs is to group a set of header files required by AOTInductor Standalone in a separate directory, ensuring they are implemented in a header-only manner. More specifically, here is what this PR does:
* Extract the DeviceType enum class into a standalone header file in a new torch/standalone/header_only directory
* Retain the existing c10/core/DeviceType.[h|cpp] files to handle complex logic and static variables
* Import symbols from the new torch::standalone namespace into c10 for backward compatibility

This is an updated version of https://github.com/pytorch/pytorch/pull/152787, because we need to land in fbcode first. See the original comments and discussions in https://github.com/pytorch/pytorch/pull/152787.

Test Plan: CI

Reviewed By: janeyx99

Differential Revision: D75313723

desertfire commented:
Unlanded internally. Let's not force-merge here.
